File: ajustable_tuner/lib/audio_processing/yin_pitch_detector.dart
import 'dart:math';
import 'dart:typed_data';

class YinPitchDetector {
  final int bufferSize;
  final int sampleRate;
  final double threshold;
  final double minFrequency;
  final double maxFrequency;
  final int _maxSearchRange;

  late Int16List _audioPcmBuffer;
  final Float32List _hannWindowBuffer;
  final Float32List _yinBuffer;

  YinPitchDetector({
    this.bufferSize = 9600,
    this.sampleRate = 48000,
    this.minFrequency = 32.7,
    this.maxFrequency = 2093.0,
    this.threshold = 0.1,
  })  : _hannWindowBuffer = Float32List(bufferSize),
        _maxSearchRange = sampleRate ~/ minFrequency,
        _yinBuffer = Float32List(sampleRate ~/ minFrequency) {
    for (var i = 0; i < bufferSize; i++) {
      _hannWindowBuffer[i] = (0.5 - 0.5 * cos(2.0 * pi * i / (bufferSize - 1)));
    }
  }

  double detectPitch(Int16List audioPcmArray) {
    _audioPcmBuffer = audioPcmArray;

    _yinBuffer[0] = 1.0;
    var runningSum = 0.0;
    var tauEstimate = -1;
    var crossThreshold = false;

    for (var tau = 1; tau < _maxSearchRange; tau++) {
      _yinBuffer[tau] = 0.0;
      for (var j = 0; j < bufferSize ~/ 2; j++) {
        var delta =
            _getModulatedAudioValue(j) - _getModulatedAudioValue(j + tau);
        _yinBuffer[tau] += delta * delta;
      }
      runningSum += _yinBuffer[tau];
      _yinBuffer[tau] *= tau / runningSum;

      if (crossThreshold) {
        if (_yinBuffer[tau - 1] <= _yinBuffer[tau]) {
          tauEstimate = tau - 1;
          break;
        }
      } else if (_yinBuffer[tau] < threshold) {
        crossThreshold = true;
      }
    }

    if (tauEstimate != -1) {
      var frequency = sampleRate / _parabolicInterpolation(tauEstimate);
      return frequency <= maxFrequency ? frequency : -1.0;
    }
    return -1.0;
  }

  double _getModulatedAudioValue(int j) {
    return _audioPcmBuffer[j] * _hannWindowBuffer[j];
  }

  double _parabolicInterpolation(int tauEstimate) {
    var x0 = (tauEstimate < 1) ? tauEstimate : tauEstimate - 1;
    var x2 =
        (tauEstimate + 1 < _maxSearchRange) ? tauEstimate + 1 : tauEstimate;

    if (x0 == tauEstimate) {
      return (_yinBuffer[tauEstimate] <= _yinBuffer[x2]) ? 0 : x2.toDouble();
    } else if (x2 == tauEstimate) {
      return (_yinBuffer[tauEstimate] <= _yinBuffer[x0])
          ? tauEstimate.toDouble()
          : x0.toDouble();
    } else {
      var s0 = _yinBuffer[x0];
      var s1 = _yinBuffer[tauEstimate];
      var s2 = _yinBuffer[x2];
      return tauEstimate + (s2 - s0) / (2 * (2 * s1 - s2 - s0));
    }
  }
}

File: ajustable_tuner/android/app/src/main/kotlin/com/matheusdev/adjustable_tuner/audio/YinPitchDetector.kt
package com.matheusdev.adjustable_tuner.audio

import kotlin.math.cos

/**
 * A pitch detection implementation using the YIN algorithm.
 *
 * @property bufferSize The size of the buffer used for processing the audio signal.
 * @property sampleRate The rate at which audio samples are taken.
 * @property threshold The threshold used in the YIN algorithm to detect pitch.
 * @constructor Initializes the detector with the specified buffer size, sample rate, and threshold.
 */
class YinPitchDetector(
    private val bufferSize: Int,
    private val sampleRate: Int,
    private val threshold: Float = DEFAULT_YIN_THRESHOLD
) {
    private lateinit var audioPcmBuffer: ShortArray

    // Variables optimized for memory usage due to frequent access
    private val hannWindowBuffer: FloatArray = FloatArray(bufferSize)
    private val maxSearchRange = (sampleRate / MIN_FREQUENCY).toInt()
    private val yinBuffer = FloatArray(maxSearchRange)

    init {
        for (i in hannWindowBuffer.indices) {
            hannWindowBuffer[i] = (0.5 - 0.5 * cos(2.0 * Math.PI * i / (bufferSize - 1))).toFloat()
        }
    }

    /**
     * Detects the pitch of a given audio sample.
     *
     * @param audioPcmArray The audio sample used for pitch detection.
     * @return The detected pitch in Hz or -1.0 if no pitch is detected.
     */
    fun detectPitch(audioPcmArray: ShortArray): Double {
        this.audioPcmBuffer = audioPcmArray

        yinBuffer[0] = 1.0f
        var runningSum = 0.0f
        var tauEstimate: Int = -1
        var crossThreshold = false

        for (tau in 1 until maxSearchRange) {
            //computeDifference
            yinBuffer[tau] = 0.0f
            for (j in 0 until bufferSize / 2) {
                val delta = getModulatedAudioValue(j) - getModulatedAudioValue(j + tau)
                yinBuffer[tau] += delta * delta
            }

            // Cumulative mean normalized difference function
            runningSum += yinBuffer[tau]
            yinBuffer[tau] *= tau / runningSum

            // Absolute threshold
            if (crossThreshold) {
                if (yinBuffer[tau - 1] <= yinBuffer[tau]) {
                    tauEstimate = tau - 1
                    break
                }
            } else if (yinBuffer[tau] < threshold) {
                crossThreshold = true
            }
        }

        return if (tauEstimate != -1) {
            val frequency = sampleRate / parabolicInterpolation(tauEstimate)
            if (frequency <= MAX_FREQUENCY) frequency.toDouble() else -1.0
        } else {
            -1.0
        }
    }

    /**
     * Hann window to the input signal which can improve the frequency estimation
     * by reducing spectral leakage
     */
    private fun getModulatedAudioValue(j: Int): Float =
        audioPcmBuffer[j] * hannWindowBuffer[j]

    /**
     * By using parabolic interpolation, you can obtain a more accurate estimate
     * that takes into account values between discrete samples.
     */
    private fun parabolicInterpolation(tauEstimate: Int): Float {
        val x0: Int = if (tauEstimate < 1) tauEstimate else tauEstimate - 1
        val x2: Int = if (tauEstimate + 1 < maxSearchRange) tauEstimate + 1 else tauEstimate

        return when {
            x0 == tauEstimate -> {
                (if (yinBuffer[tauEstimate] <= yinBuffer[x2]) 0 else x2).toFloat()
            }
            x2 == tauEstimate -> {
                (if (yinBuffer[tauEstimate] <= yinBuffer[x0]) tauEstimate else x0).toFloat()
            }
            else -> {
                val s0 = yinBuffer[x0]
                val s1 = yinBuffer[tauEstimate]
                val s2 = yinBuffer[x2]
                tauEstimate + (s2 - s0) / (2 * (2 * s1 - s2 - s0))
            }
        }
    }

    companion object {
        private const val DEFAULT_YIN_THRESHOLD = 0.07f
        /**
         * Frequencies on string instruments spectrum
         */
        private const val MIN_FREQUENCY = 32.7f
        private const val MAX_FREQUENCY = 2093.0f
    }
}
